source/credentials_provider_sts.c

```diff
@@ -0,0 +1,668 @@
+    (void)error_code;
+    struct sts_creds_provider_user_data *provider_user_data = userdata;
+    struct aws_credentials_provider_sts_impl *sts_impl = provider_user_data->provider->impl;
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+        "(id=%p): signing completed with error code %d",
+        (void *)provider_user_data->provider,
+        error_code);
+
+    if (error_code) {
+        aws_raise_error(error_code);
+        goto error;
+    }
+
+    if (aws_apply_signing_result_to_http_request(
+            provider_user_data->message, provider_user_data->provider->allocator, result)) {
+        goto error;
+    }
+
+    sts_impl->function_table->aws_http_connection_manager_acquire_connection(
+        sts_impl->connection_manager, s_on_connection_setup_fn, provider_user_data);
+    return;
+
+error:
+    provider_user_data->callback(NULL, provider_user_data->user_data);
+    s_clean_up_user_data(provider_user_data);
+}
+
+static int s_sts_get_creds(
+    struct aws_credentials_provider *provider,
+    aws_on_get_credentials_callback_fn callback,
+    void *user_data) {
+
+    AWS_LOGF_DEBUG(AWS_LS_AUTH_CREDENTIALS_PROVIDER, "(id=%p): fetching credentials", (void *)provider);
+
+    struct aws_credentials_provider_sts_impl *sts_impl = provider->impl;
+    struct sts_creds_provider_user_data *provider_user_data =
+        aws_mem_calloc(provider->allocator, 1, sizeof(struct sts_creds_provider_user_data));
+
+    if (!provider_user_data) {
+        goto error;
+    }
+
+    provider_user_data->allocator = provider->allocator;
+    provider_user_data->provider = provider;
+    provider_user_data->callback = callback;
+    provider_user_data->user_data = user_data;
+
+    provider_user_data->message = aws_http_message_new_request(provider->allocator);
+
+    if (!provider_user_data->message) {
+        goto error;
+    }
+
+    struct aws_http_header host_header = {
+        .name = s_host_name,
+        .value = s_host_name_val,
+    };
+
+    if (aws_http_message_add_header(provider_user_data->message, host_header)) {
+        goto error;
+    }
+
+    struct aws_http_header content_type_header = {
+        .name = s_content_type,
+        .value = s_content_type_val,
+    };
+
+    if (aws_http_message_add_header(provider_user_data->message, content_type_header)) {
+        goto error;
+    }
+
+    struct aws_http_header api_version_header = {
+        .name = s_api_version,
+        .value = s_api_version_val,
+    };
+
+    if (aws_http_message_add_header(provider_user_data->message, api_version_header)) {
+        goto error;
+    }
+
+    if (aws_byte_buf_init(&provider_user_data->payload_body, provider->allocator, 256)) {
+        goto error;
+    }
+
+    if (s_write_body_to_buffer(provider, &provider_user_data->payload_body)) {
+        goto error;
+    }
+
+    char content_length[21];
+    AWS_ZERO_ARRAY(content_length);
+    sprintf(content_length, "%" PRIu64, (uint64_t)provider_user_data->payload_body.len);
```

Comment:
I know it's safe, but would prefer snprintf for these types of things.

Reply:
snprintf is not standard.... also modern compilers check it. I'd be up for adding a standard one to aws-c-common though.

Reply:
snprintf is standard though.  it's sprintf_s that isn't.

Reply:
done